File: internal/context/context.go
package context

import (
	"bufio"
	"bytes"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"text/template"
	"time"

	"github.com/delightfulhammers/telesis/internal/config"
	"github.com/delightfulhammers/telesis/templates"
)

type templateData struct {
	ProjectName       string
	ProjectOwner      string
	ProjectLanguage   string
	ProjectStatus     string
	ProjectRepo       string
	GeneratedDate     string
	Description       string
	MilestonesContent string
	ADRs              []string
	ADRCount          int
	TDDCount          int
	Principles        string
	ContextSections   []contextSection
}

type contextSection struct {
	Content string
}

func Generate(rootDir string) (string, error) {
	cfg, err := config.Load(rootDir)
	if err != nil {
		return "", err
	}

	data := templateData{
		ProjectName:     cfg.Project.Name,
		ProjectOwner:    cfg.Project.Owner,
		ProjectLanguage: cfg.Project.Language,
		ProjectStatus:   cfg.Project.Status,
		ProjectRepo:     cfg.Project.Repo,
		GeneratedDate:   time.Now().Format("2006-01-02"),
	}

	adrs, adrCount, err := scanADRs(filepath.Join(rootDir, "docs", "adr"))
	if err != nil {
		return "", fmt.Errorf("scanning ADRs: %w", err)
	}
	data.ADRs = adrs
	data.ADRCount = adrCount

	tddCount, err := countFiles(filepath.Join(rootDir, "docs", "tdd"), "TDD-*.md")
	if err != nil {
		return "", fmt.Errorf("counting TDDs: %w", err)
	}
	data.TDDCount = tddCount

	milestones, err := extractMilestones(filepath.Join(rootDir, "docs", "MILESTONES.md"))
	if err != nil {
		return "", fmt.Errorf("reading milestones: %w", err)
	}
	data.MilestonesContent = milestones

	principles, err := extractPrinciples(filepath.Join(rootDir, "docs", "VISION.md"))
	if err != nil {
		return "", fmt.Errorf("reading vision: %w", err)
	}
	data.Principles = principles

	description, err := extractSection(filepath.Join(rootDir, "docs", "VISION.md"), descriptionHeaderRe)
	if err != nil {
		return "", fmt.Errorf("reading description: %w", err)
	}
	data.Description = stripLeadingHeading(description)

	contextSections, err := scanContextFiles(filepath.Join(rootDir, "docs", "context"))
	if err != nil {
		return "", fmt.Errorf("reading context files: %w", err)
	}
	data.ContextSections = contextSections

	tmplContent, err := templates.FS.ReadFile("claude.md.tmpl")
	if err != nil {
		return "", fmt.Errorf("could not read template: %w", err)
	}

	tmpl, err := template.New("claude").Parse(string(tmplContent))
	if err != nil {
		return "", fmt.Errorf("could not parse template: %w", err)
	}

	var buf bytes.Buffer
	if err := tmpl.Execute(&buf, data); err != nil {
		return "", fmt.Errorf("could not render template: %w", err)
	}

	return buf.String(), nil
}

var adrNumberRe = regexp.MustCompile(`^ADR-(\d+)`)

type numberedADR struct {
	path string
	num  int
}

func scanADRs(adrDir string) ([]string, int, error) {
	entries, err := os.ReadDir(adrDir)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, 0, nil
		}
		return nil, 0, fmt.Errorf("reading ADR directory: %w", err)
	}

	var adrs []numberedADR
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		matched, _ := filepath.Match("ADR-*.md", entry.Name())
		if matched {
			p := filepath.Join(adrDir, entry.Name())
			adrs = append(adrs, numberedADR{path: p, num: parseADRNumber(p)})
		}
	}
	if len(adrs) == 0 {
		return nil, 0, nil
	}

	sort.Slice(adrs, func(i, j int) bool {
		return adrs[i].num < adrs[j].num
	})

	// Return up to 5 most recent (highest numbered)
	start := 0
	if len(adrs) > 5 {
		start = len(adrs) - 5
	}
	recent := adrs[start:]

	var summaries []string
	for _, adr := range recent {
		summary, err := extractADRSummary(adr.path)
		if err != nil {
			return nil, 0, fmt.Errorf("reading ADR %s: %w", filepath.Base(adr.path), err)
		}
		summaries = append(summaries, summary)
	}

	return summaries, len(adrs), nil
}

func parseADRNumber(path string) int {
	name := filepath.Base(path)
	if m := adrNumberRe.FindStringSubmatch(name); len(m) > 1 {
		n, _ := strconv.Atoi(m[1])
		return n
	}
	return 0
}

var adrTitleRe = regexp.MustCompile(`^#\s+ADR-\d+:\s*(.+)`)

func extractADRSummary(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()

	filename := filepath.Base(path)
	name := strings.TrimSuffix(filename, ".md")

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if m := adrTitleRe.FindStringSubmatch(line); len(m) > 1 {
			return fmt.Sprintf("%s: %s", name, m[1]), nil
		}
	}
	if err := scanner.Err(); err != nil {
		return "", fmt.Errorf("scanning %s: %w", filename, err)
	}

	return name, nil
}

func countFiles(dir, pattern string) (int, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return 0, nil
		}
		return 0, fmt.Errorf("reading %s: %w", dir, err)
	}
	count := 0
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		matched, _ := filepath.Match(pattern, entry.Name())
		if matched {
			count++
		}
	}
	return count, nil
}

var (
	principlesHeaderRe  = regexp.MustCompile(`(?i)^##\s+Design Principles`)
	descriptionHeaderRe = regexp.MustCompile(`(?i)^##\s+The Vision`)
)

func extractPrinciples(visionPath string) (string, error) {
	f, err := os.Open(visionPath)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return "", nil
		}
		return "", err
	}
	defer f.Close()

	var capturing bool
	var result []string

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		if principlesHeaderRe.MatchString(line) {
			capturing = true
			continue
		}
		if capturing {
			if strings.HasPrefix(line, "## ") || strings.HasPrefix(line, "---") {
				break
			}
			result = append(result, line)
		}
	}
	if err := scanner.Err(); err != nil {
		return "", fmt.Errorf("scanning vision: %w", err)
	}

	return strings.TrimSpace(strings.Join(result, "\n")), nil
}

var milestonesHeaderRe = regexp.MustCompile(`(?i)^##\s+MVP`)

func extractMilestones(milestonesPath string) (string, error) {
	return extractSection(milestonesPath, milestonesHeaderRe)
}

// extractSection extracts content under the first heading matching the given
// compiled regex, stopping at the next ## heading or --- separator. The matched
// heading line is included in the output.
func extractSection(path string, re *regexp.Regexp) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return "", nil
		}
		return "", err
	}
	defer f.Close()

	var capturing bool
	var result []string

	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		if !capturing && re.MatchString(line) {
			capturing = true
			result = append(result, line)
			continue
		}
		if capturing {
			trimmed := strings.TrimSpace(line)
			if strings.HasPrefix(trimmed, "## ") || trimmed == "---" {
				break
			}
			result = append(result, line)
		}
	}
	if err := scanner.Err(); err != nil {
		return "", fmt.Errorf("scanning %s: %w", filepath.Base(path), err)
	}

	return strings.TrimSpace(strings.Join(result, "\n")), nil
}

// stripLeadingHeading removes a markdown heading from the first line if present,
// returning just the body content. Used when the template provides its own heading.
func stripLeadingHeading(s string) string {
	lines := strings.SplitN(s, "\n", 2)
	if len(lines) > 0 && strings.HasPrefix(strings.TrimSpace(lines[0]), "#") {
		if len(lines) > 1 {
			return strings.TrimSpace(lines[1])
		}
		return ""
	}
	return s
}

func scanContextFiles(contextDir string) ([]contextSection, error) {
	entries, err := os.ReadDir(contextDir)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, err
	}

	var sections []contextSection
	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".md") {
			continue
		}
		content, err := os.ReadFile(filepath.Join(contextDir, entry.Name()))
		if err != nil {
			return nil, fmt.Errorf("reading %s: %w", entry.Name(), err)
		}
		sections = append(sections, contextSection{
			Content: strings.TrimSpace(string(content)),
		})
	}
	return sections, nil
}

File: src/context/context.ts
import {
  readFileSync,
  readdirSync,
  writeFileSync,
  renameSync,
  unlinkSync,
  openSync,
  closeSync,
  constants,
  type Dirent,
} from "node:fs";
import { join, basename } from "node:path";
import { load } from "../config/config.js";
import type { DocLayer, DocLayerScope } from "../config/config.js";
import { extractActiveMilestone } from "../milestones/parse.js";
import { renderTemplate } from "../templates/index.js";
import { loadNotes } from "../notes/store.js";
import { renderNotesSection } from "../notes/format.js";
import { loadEntries } from "../journal/store.js";
import { renderJournalSection } from "../journal/format.js";

interface ContextSection {
  readonly Content: string;
}

interface NumberedADR {
  readonly path: string;
  readonly num: number;
}

const ADR_NUMBER_RE = /^ADR-(\d+)/;
const ADR_TITLE_RE = /^#\s+ADR-\d+:\s*(.+)/;
const PRINCIPLES_HEADER_RE = /^##\s+Design Principles/i;
const DESCRIPTION_HEADER_RE = /^##\s+The Vision/i;

const isENOENT = (err: unknown): boolean =>
  (err as NodeJS.ErrnoException).code === "ENOENT";

const isSectionBoundary = (trimmed: string): boolean => {
  if (trimmed === "---") return true;
  if (trimmed.startsWith("# ") && !trimmed.startsWith("##")) return true;
  if (trimmed.startsWith("## ")) return true;
  return false;
};

const extractSection = (
  path: string,
  re: RegExp,
  includeHeading: boolean,
): string => {
  let content: string;
  try {
    content = readFileSync(path, "utf-8");
  } catch (err) {
    if (isENOENT(err)) return "";
    throw err;
  }
  const lines = content.split("\n");

  let capturing = false;
  const result: string[] = [];

  for (const line of lines) {
    if (!capturing && re.test(line)) {
      capturing = true;
      if (includeHeading) {
        result.push(line);
      }
      continue;
    }
    if (capturing) {
      const trimmed = line.trim();
      if (isSectionBoundary(trimmed)) {
        break;
      }
      result.push(line);
    }
  }

  return result.join("\n").trim();
};

const parseADRNumber = (path: string): number => {
  const name = basename(path);
  const m = ADR_NUMBER_RE.exec(name);
  if (m && m[1]) {
    return parseInt(m[1], 10);
  }
  return 0;
};

const extractADRSummary = (path: string): string => {
  const content = readFileSync(path, "utf-8");
  const filename = basename(path);
  const name = filename.replace(/\.md$/, "");

  const lines = content.split("\n");
  for (const line of lines) {
    const trimmed = line.trim();
    const m = ADR_TITLE_RE.exec(trimmed);
    if (m && m[1]) {
      return `${name}: ${m[1]}`;
    }
  }

  return name;
};

const scanADRs = (adrDir: string): { summaries: string[]; count: number } => {
  let entries: Dirent[];
  try {
    entries = readdirSync(adrDir, { withFileTypes: true });
  } catch (err) {
    if (isENOENT(err)) return { summaries: [], count: 0 };
    throw err;
  }

  const adrs: NumberedADR[] = entries
    .filter((entry) => !entry.isDirectory() && /^ADR-.*\.md$/.test(entry.name))
    .map((entry) => ({
      path: join(adrDir, entry.name),
      num: parseADRNumber(entry.name),
    }));

  if (adrs.length === 0) return { summaries: [], count: 0 };

  adrs.sort((a, b) => a.num - b.num);

  // Return up to 5 most recent (highest numbered)
  const start = Math.max(0, adrs.length - 5);
  const recent = adrs.slice(start);

  const summaries = recent.map((adr) => extractADRSummary(adr.path));

  return { summaries, count: adrs.length };
};

const countFiles = (dir: string, pattern: RegExp): number => {
  let entries: Dirent[];
  try {
    entries = readdirSync(dir, { withFileTypes: true });
  } catch (err) {
    if (isENOENT(err)) return 0;
    throw err;
  }

  return entries.filter(
    (entry) => !entry.isDirectory() && pattern.test(entry.name),
  ).length;
};

interface ScannedTDD {
  readonly name: string;
  readonly status: string;
  readonly overview: string;
  readonly interfaces: string;
  readonly path: string;
  readonly num: number;
}

const TDD_NUMBER_RE = /^TDD-(\d+)/;
const TDD_TITLE_RE = /^#\s+TDD-\d+\s*[—–-]\s*(.+)/m;
const TDD_STATUS_RE = /^\*\*Status:\*\*\s*(.+)/m;
const OVERVIEW_HEADER_RE = /^##\s+Overview/i;
const INTERFACES_HEADER_RE = /^##\s+Interfaces/i;

const MAX_INLINED_TDDS = 10;

/** Extract a section from content string (avoids re-reading the file). */
const extractSectionFromContent = (content: string, re: RegExp): string => {
  const lines = content.split("\n");
  let capturing = false;
  const result: string[] = [];

  for (const line of lines) {
    if (!capturing && re.test(line)) {
      capturing = true;
      continue;
    }
    if (capturing) {
      const trimmed = line.trim();
      if (isSectionBoundary(trimmed)) break;
      result.push(line);
    }
  }

  return result.join("\n").trim();
};

const scanTDDs = (tddDir: string): { items: ScannedTDD[]; count: number } => {
  let entries: Dirent[];
  try {
    entries = readdirSync(tddDir, { withFileTypes: true });
  } catch (err) {
    if (isENOENT(err)) return { items: [], count: 0 };
    throw err;
  }

  const tddFiles = entries.filter(
    (entry) => !entry.isDirectory() && /^TDD-.*\.md$/.test(entry.name),
  );

  if (tddFiles.length === 0) return { items: [], count: 0 };

  const tdds: ScannedTDD[] = [];

  for (const entry of tddFiles) {
    const filePath = join(tddDir, entry.name);
    const content = readFileSync(filePath, "utf-8");

    // Extract status
    let status = "Draft";
    const statusMatch = TDD_STATUS_RE.exec(content);
    if (statusMatch?.[1]) {
      status = statusMatch[1].trim();
    }

    // Skip superseded TDDs
    if (status.toLowerCase() === "superseded") continue;

    // Extract TDD number
    const numMatch = TDD_NUMBER_RE.exec(entry.name);
    const num = numMatch?.[1] ? parseInt(numMatch[1], 10) : 0;

    // Extract title
    let name = entry.name.replace(/\.md$/, "");
    const titleMatch = TDD_TITLE_RE.exec(content);
    if (titleMatch?.[1]) {
      name = `${entry.name.replace(/\.md$/, "").split("-").slice(0, 2).join("-")}: ${titleMatch[1].trim()}`;
    }

    // Extract sections from the single read — no re-reading the file
    const overview = extractSectionFromContent(content, OVERVIEW_HEADER_RE);
    const interfaces = extractSectionFromContent(content, INTERFACES_HEADER_RE);

    tdds.push({
      name,
      status,
      overview,
      interfaces,
      path: entry.name,
      num,
    });
  }

  // Sort by number, take most recent
  tdds.sort((a, b) => a.num - b.num);
  const recent = tdds.slice(Math.max(0, tdds.length - MAX_INLINED_TDDS));

  return { items: recent, count: tddFiles.length };
};

const scanContextFiles = (contextDir: string): ContextSection[] => {
  let entries: Dirent[];
  try {
    entries = readdirSync(contextDir, { withFileTypes: true });
  } catch (err) {
    if (isENOENT(err)) return [];
    throw err;
  }

  return entries
    .filter((entry) => !entry.isDirectory() && entry.name.endsWith(".md"))
    .map((entry) => entry.name)
    .sort()
    .map((name) => ({
      Content: readFileSync(join(contextDir, name), "utf-8").trim(),
    }));
};

/** Check if a layer's include list covers a given scope */
const layerIncludes = (layer: DocLayer, scope: DocLayerScope): boolean => {
  return layer.include.includes("all") || layer.include.includes(scope);
};

export const generate = (rootDir: string): string => {
  const cfg = load(rootDir);
  const contextConfig = cfg.context ?? {
    layers: [{ path: "docs", include: ["all" as DocLayerScope] }],
  };

  // Collect ADRs, TDDs, and context files across all layers
  let allADRSummaries: string[] = [];
  let totalADRCount = 0;
  let allTDDItems: ScannedTDD[] = [];
  let totalTDDCount = 0;
  let allContextSections: ContextSection[] = [];

  // For singular docs, use the last (most local) layer that includes them
  let milestonesContent = "";
  let principles = "";
  let description = "";

  for (const layer of contextConfig.layers) {
    const layerPath = join(rootDir, layer.path);

    if (layerIncludes(layer, "adrs")) {
      const { summaries, count } = scanADRs(join(layerPath, "adr"));
      allADRSummaries = allADRSummaries.concat(summaries);
      totalADRCount += count;
    }

    if (layerIncludes(layer, "tdds")) {
      const { items: layerTDDs, count: layerTDDCount } = scanTDDs(
        join(layerPath, "tdd"),
      );
      allTDDItems = allTDDItems.concat(layerTDDs);
      totalTDDCount += layerTDDCount;
    }

    if (layerIncludes(layer, "context")) {
      const sections = scanContextFiles(join(layerPath, "context"));
      allContextSections = allContextSections.concat(sections);
    }

    if (layerIncludes(layer, "milestones")) {
      const content = extractActiveMilestone(join(layerPath, "MILESTONES.md"));
      if (content) milestonesContent = content;
    }

    if (layerIncludes(layer, "vision")) {
      const p = extractSection(
        join(layerPath, "VISION.md"),
        PRINCIPLES_HEADER_RE,
        false,
      );
      if (p) principles = p;

      const d = extractSection(
        join(layerPath, "VISION.md"),
        DESCRIPTION_HEADER_RE,
        false,
      );
      if (d) description = d;
    }
  }

  // Deduplicate ADR summaries by number (later layers override), sort, take 5.
  // Unparseable ADRs (num=0) get unique negative keys to avoid collision.
  const adrByNum = new Map<number, string>();
  let unparseableIdx = 0;
  for (const summary of allADRSummaries) {
    const num = parseADRNumber(summary);
    const key = num === 0 ? -(++unparseableIdx) : num;
    adrByNum.set(key, summary);
  }
  const sortedADRs = [...adrByNum.values()].sort((a, b) => {
    return parseADRNumber(a) - parseADRNumber(b);
  });
  const recentADRs = sortedADRs.slice(Math.max(0, sortedADRs.length - 5));

  // Deduplicate TDDs by num, sort, cap at MAX_INLINED_TDDS.
  // Unparseable TDDs (num=0) get unique negative keys to avoid collision.
  const tddByNum = new Map<number, ScannedTDD>();
  let unparseableTDDIdx = 0;
  for (const tdd of allTDDItems) {
    const key = tdd.num === 0 ? -(++unparseableTDDIdx) : tdd.num;
    tddByNum.set(key, tdd);
  }
  const sortedTDDs = [...tddByNum.values()].sort((a, b) => a.num - b.num);
  const recentTDDs = sortedTDDs.slice(
    Math.max(0, sortedTDDs.length - MAX_INLINED_TDDS),
  );

  const { items: notes } = loadNotes(rootDir);
  const notesSection = renderNotesSection(notes);

  const { items: journalEntries } = loadEntries(rootDir);
  const journalSection = renderJournalSection(journalEntries);

  const now = new Date();
  const generatedDate = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}-${String(now.getDate()).padStart(2, "0")}`;

  return renderTemplate("claude.md.tmpl", {
    ProjectName: cfg.project.name,
    ProjectOwner: cfg.project.owner,
    ProjectLanguage: cfg.project.languages.join(", ") || cfg.project.language,
    ProjectStatus: cfg.project.status,
    ProjectRepo: cfg.project.repo,
    GeneratedDate: generatedDate,
    Description: description,
    MilestonesContent: milestonesContent,
    ADRs: recentADRs.length > 0 ? { items: recentADRs } : false,
    ADRCount: totalADRCount,
    TDDCount: totalTDDCount,
    TDDs: recentTDDs.length > 0 ? { items: recentTDDs } : false,
    Principles: principles,
    ContextSections: allContextSections,
    JournalSection: journalSection || false,
    NotesSection: notesSection || false,
  });
};

/**
 * Generates CLAUDE.md and atomically writes it to the project root.
 * Uses temp file + rename to avoid partial writes.
 */
export const generateAndWrite = (rootDir: string): void => {
  const output = generate(rootDir);
  const claudePath = join(rootDir, "CLAUDE.md");
  const suffix = `${process.pid}-${Date.now()}-${Math.random().toString(36).slice(2)}`;
  const tmpPath = join(rootDir, `.CLAUDE-${suffix}.md`);

  const fd = openSync(
    tmpPath,
    constants.O_WRONLY | constants.O_CREAT | constants.O_EXCL,
    0o666,
  );

  try {
    writeFileSync(fd, output);
  } catch (err) {
    try {
      closeSync(fd);
    } catch {
      /* best-effort */
    }
    try {
      unlinkSync(tmpPath);
    } catch {
      /* best-effort */
    }
    throw err;
  }

  closeSync(fd);

  try {
    renameSync(tmpPath, claudePath);
  } catch (err) {
    try {
      unlinkSync(tmpPath);
    } catch {
      /* cleanup best-effort */
    }
    throw err;
  }
};
